### PR TITLE
ipam: introducing contivCIDR to calculate pod/node/vxlan address spaces

### PR DESCRIPTION
**Pull request by brecode**
This PR removes the ability from the user to define pod/node/vxlan address space.

  * Introducing ContivCIDR in the contiv-vpp.yaml from which, all the subnets are being calculated
    A /14 prefix (or less, i.e /13, /12) is needed. 
  * From the given CIDR we get two subnets of 32768 (or more) addresses each assigned for 
    PodSubnetCIDR and VPPHostSubnetCIDR
  * Three /23 subnets are reserved for NodeInterconnectCIDR, VxlanCIDR and podIfIPCIDR
  * Modified Vagrant to reflect changes 
  * Modified Helm charts
  * Modified yaml (k8s/contiv-vpp + arm)

### CODE REVIEW

**Comment by jmedved** [comment]
"Three /24 subnets are reserved for NodeInterconnectCIDR, VxlanCIDR and podIfIPCIDR" - this will mean we can only have up to 256 nodes. we need a <=23 subnets for all these to support 500 nodes.

**Comment by brecode** [comment]
Yup, before I change it to /24 I wanted to consult with the @rastislavszabo and @lukasmacko if the nodeID handles more than 256.

File: plugins/contiv/plugin_impl_contiv.go [hunk 1]
```diff
@@ -620,3 +623,100 @@ func appendIfMissing(slice []string, s string) []string {
...
+	return config
+}
+
+// subnet takes a CIDR range and creates a subnet from it
+// base: parent CIDR range
+// newBits: number of additional prefix bits
+// num: given network number.
+//
+// Example: 10.1.0.0/16, with additional 8 bits and a network number of 5
+// result = 10.3.5.0/24
```
Comment (lukasmacko) [inline plugins/contiv/plugin_impl_contiv.go]: result should be `10.1.5.0/24` shoudln't it?

Comment (brecode) [reply plugins/contiv/plugin_impl_contiv.go]: yup typo

**Comment by lukasmacko** [comment]
Could you add a simle unit test that will show how the default ContivCIDR is split?

Correct me if I'm wrong however it seems to me that there is an overlap in generated subnets:
Given `10.1.0.0/16` as ContivCIDR

| name   | subnet    |
|---|---|
| podSubnetCIDR  | 10.1.0.0/17 | 
| vppHostSubnetCIDR  | **10.1.128.0/17** includes all three below |
| nodeInterconnectCIDR  | 10.1.250.0/23 |
| podIfIPCIDR  | 10.1.252.0/23 |
| vxlanCIDR  | 10.1.254.0/23 |

**Comment by brecode** [comment]
@lukasmacko You're right, I thought we could handle address overlapping although it's not good practise, so we don't go into a bigger network and have so many unused addresses. 

I've now increased the ContivCIDR to /14 so we can have ~ 65K pods in the cluster and 512 nodes but we're left with a many unused addresses.

File: plugins/contiv/plugin_impl_contiv.go [hunk 2]
```diff
@@ -626,17 +626,17 @@ func appendIfMissing(slice []string, s string) []string {
 
 // getIPAMConfig populates the Config struct with the calculated subnets
 func getIPAMConfig(config *Config) *Config {
-	_, contivNetwork, _ := net.ParseCIDR(config.IPAMConfig.ContivCIDR)
+	_, contivNetwork, _ := net.ParseCIDR("10.1.0.0/14")
```
Comment (lukasmacko) [inline plugins/contiv/plugin_impl_contiv.go]: the param should be used instead of constant

**Comment by jmedved** [comment]
One more thing occurred to me: we should provide an option to configure NodeInterconnectCIDR separately from the rest of the IP address ranges, since a user may want to allocate their own visible address range for that (e.g. from their own public address space)

**Comment by brecode** [comment]
@jmedved so are we leaving the nodeInterconnectCIDR for the user to decide ?

File: k8s/contiv-vpp/values.yaml [hunk 3]
```diff
@@ -14,7 +14,7 @@ contiv:
   serviceLocalEndpointWeight: 1
   disableNATVirtualReassembly: True
   ipamConfig:
-    contivCIDR: "10.1.0.0/16"
+    contivCIDR: "10.1.0.0/15"
```
Comment (lukasmacko) [inline k8s/contiv-vpp/values.yaml]: `.1` the second octet will be ingnored due to mask used, rather use `10.2.0.0/15`

File: plugins/contiv/plugin_impl_contiv.go (hunk 2 above)
Comment (brecode) [reply plugins/contiv/plugin_impl_contiv.go]: Right, this went in from my testing.

File: k8s/contiv-vpp/values.yaml (hunk 3 above)
Comment (brecode) [reply k8s/contiv-vpp/values.yaml]: I think this needs to change to 10.0.0.0/14 now, right?

**Comment by rastislavszabo** [comment]
> One more thing occurred to me: we should provide an option to configure NodeInterconnectCIDR separately from the rest of the IP address ranges, since a user may want to allocate their own visible address range for that (e.g. from their own public address space)

+1 for this

I was even thinking that we might want to still allow to setup all individual subsets by the user as before (instead of ContivCIDR), at least as a "secret" option. Just in case that someone would come in later and tell us they cannot afford so large subnet.

File: k8s/contiv-vpp/values.yaml (hunk 3 above)
Comment (lukasmacko) [reply k8s/contiv-vpp/values.yaml]: yes that's another option

**Comment by brecode** [comment]
+1 from me, we actually encountered something similar w/ @jmedved at a debugging session today. We can keep the NodeInterconnectCIDR field and either use the large subnet if it's empty or split the subnets differently if exists in the contiv-vpp.yaml (default value=nil, if introduced by the user we use that instead)

**Comment by brecode** [comment]
Closing this, implemented in #1101

**Pull request closed without merging**